src/libsyntax/ext/deriving/generic/mod.rs

```diff
@@ -534,7 +542,22 @@ impl<'a> TraitDef<'a> {
                                      body)
         }).collect();
 
-        self.create_derived_impl(cx, type_ident, generics, methods)
+        // Transform associated types from `deriving::ty::Ty` into `ast::Typedef`
+        let types = self.associated_types.iter().map(|&(ident, ref type_def)| {
```

Comment:
I think this would be better suited to being in `create_derived_impl`, since it's independent of whether the type is a struct or enum.